controllers/kubeadmconfig_controller.go

```diff
@@ -104,41 +107,107 @@ func (r *KubeadmConfigReconciler) Reconcile(req ctrl.Request) (ctrl.Result, erro
 	}
 
+	// Check for infrastructure ready. If it's not ready then we will requeue the machine until it is.
+	// The cluster-api machine controller set this value.
+	if cluster.Status.InfrastructureReady != true {
+		log.Info("Infrastructure is not ready, requeing until ready.")
+		return ctrl.Result{RequeueAfter: 30 * time.Second}, nil
+	}
+
 	// Store Config's state, pre-modifications, to allow patching
 	patchConfig := client.MergeFrom(config.DeepCopy())
 
-	// if the machine has cluster and or init defined on it then generate the init regular join
-	if config.Spec.InitConfiguration != nil && config.Spec.ClusterConfiguration != nil {
-		// get both of these to strings to pass to the cloud init control plane generator
+	// Check for control plane ready. If it's not ready then we will requeue the machine until it is.
+	// The cluster-api machine controller set this value.
+	if cluster.Annotations[ControlPlaneReadyAnnotationKey] != "true" {
+		// if it's NOT a control plane machine, requeue
+		if !util.IsControlPlaneMachine(machine) {
+			log.Info("Control plane is not ready, requeing worker nodes until ready.")
+			return ctrl.Result{RequeueAfter: 30 * time.Second}, nil
+		}
+
+		// if the machine has not ClusterConfiguration and InitConfiguration, requeue
+		if config.Spec.InitConfiguration == nil && config.Spec.ClusterConfiguration == nil {
+			log.Info("Control plane is not ready, requeing joining control planes until ready.")
+			return ctrl.Result{RequeueAfter: 30 * time.Second}, nil
+		}
+
+		//TODO(fp) use init lock so only the first machine configured as control plane get processed, everything else gets requeued
+
+		// otherwise it is a init control plane
+		// Nb. in this case JoinConfiguration should not be defined by users, but in case of misconfigurations, CAPBK simply ignore it
+
+		log.Info("Creating BootstrapData for the init control plane")
+
+		// get both of ClusterConfiguration and InitConfiguration strings to pass to the cloud init control plane generator
+		// kubeadm allows one this values to be empty; cabpk replace missing values with an empty config, so the cloud init generation
+		// should not handle special cases.
+
+		if config.Spec.InitConfiguration == nil {
+			config.Spec.InitConfiguration = &kubeadmv1beta1.InitConfiguration{
+				TypeMeta: v1.TypeMeta{
+					APIVersion: "kubeadm.k8s.io/v1beta1",
+					Kind:       "InitConfiguration",
+				},
+			}
+		}
 		initdata, err := json.Marshal(config.Spec.InitConfiguration)
 		if err != nil {
 			log.Error(err, "failed to marshal init configuration")
 			return ctrl.Result{}, err
 		}
+
+		if config.Spec.ClusterConfiguration == nil {
+			config.Spec.ClusterConfiguration = &kubeadmv1beta1.ClusterConfiguration{
+				TypeMeta: v1.TypeMeta{
+					APIVersion: "kubeadm.k8s.io/v1beta1",
+					Kind:       "ClusterConfiguration",
+				},
+			}
+		}
+		// If there is a control plane endpoint defined at cluster in cluster status, use it as a control plane endpoint for the K8s cluster
```

Comment:
This comment should also indicate that we are only using the first one defined if there are multiple defined.